File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
// m3hu1
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
        }
        if(sum%2!=0){
            return false;
        }
        int target=sum/2;
        int index=0;
        int sum1=0;
        int n=nums.size();
        vector<vector<int>>vcc(n,vector<int>(target+1,-1));
        return Parti(nums,index,target,vcc);
    }
    bool Parti(vector<int>& nums,int index,int target,vector<vector<int>>&vcc){
        if(index>=nums.size()){
            return false;
        }
        if(target==0){
            return true;
        }
        if(target<0){
            return false;
        }
        if(vcc[index][target]!=-1){
            return vcc[index][target];
        }
        bool f=Parti(nums,index+1,target,vcc);
        bool s=Parti(nums,index+1,target-nums[index],vcc);
        return vcc[index][target]=f||s;
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.go

func canPartition(nums []int) bool {
	n := len(nums)
	SUM := 0

	for _, v := range nums { SUM += v }
	if SUM%2 == 1 { return false }

	dp := make([][]bool, n+1)
    
	for i := 0; i <= n; i++ {
		dp[i] = make([]bool, SUM)
		dp[i][0] = true
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= SUM/2; j++ {
			k := nums[i-1]
			if j-k < 0 {
				dp[i][j] = dp[i-1][j]
			} else {
				dp[i][j] = dp[i-1][j-k] || dp[i-1][j]
			}
		}
	}

	return dp[n][SUM/2]
}